### PR TITLE
✨ do not allow start service when credits bellow zero

### PR DESCRIPTION
**Pull request by matusdrobuliak66**
## What do these changes do?
- ✨ do not allow start service when credits bellow zero
image

NOTE: As discussed @bisgaard-itis we need to introduce this check also to the API-server (or start to use the same web server api to start a job) 

## Related issue/s
- https://github.com/ITISFoundation/osparc-issues/issues/922

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/director_v2/_handlers.py [hunk 1]
```diff
@@ -130,9 +131,13 @@ async def start_computation(request: web.Request) -> web.Response:
             project_wallet_id = project_wallet.wallet_id
 
         # Check whether user has access to the wallet
-        wallet = await wallets_api.get_wallet_by_user(
+        wallet = await wallets_api.get_wallet_with_available_credits_by_user_and_wallet(
             request.app, req_ctx.user_id, project_wallet_id, req_ctx.product_name
         )
+        if wallet.available_credits <= ZERO_CREDITS:
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/director_v2/_handlers.py:137]: I wonder whether zero is the right number. IMO here we should have the minimal unit that allows to run something. 
Let me argue this with an example. Think of a shop of candies. The smallest candy costs 3 dollars (it is a candy store in Zurich Banhhofstr). If your wallet has 2 dollars, you should not be able to buy anything. 
My point here is that not-enough-credits should be with that of the item with minimum price. In our case might be something as the credits to run a simulation say for 1 min??  Which is indeed not zero!

**Review by pcrespov — COMMENTED** [review]
Please consider my thought and let me know if this makes sense to you . thx!

File: services/static-webserver/client/source/class/osparc/desktop/StudyEditor.js [hunk 2]
```diff
@@ -357,6 +357,10 @@ qx.Class.define("osparc.desktop.StudyEditor", {
               this.__requestStartPipeline(studyId, partialPipeline, true);
             }
           }, this);
+        }
+        else if (e.getTarget().getStatus() == "402") {
```
Comment (odeimaiz) [inline services/static-webserver/client/source/class/osparc/desktop/StudyEditor.js:361]:
```suggestion
        } else if (e.getTarget().getStatus() == "402") {
```

File: services/static-webserver/client/source/class/osparc/data/model/Node.js [hunk 3]
```diff
@@ -967,7 +967,8 @@ qx.Class.define("osparc.data.model.Node", {
       osparc.data.Resources.fetch("studies", "startNode", params)
         .then(() => this.startDynamicService())
         .catch(err => {
-          if ("status" in err && err.status === 409) {
+          console.log(err);
```
Comment (odeimaiz) [inline services/static-webserver/client/source/class/osparc/data/model/Node.js:970]: remove that console.log 👍

**Review by odeimaiz — APPROVED** [review]

**Review by GitHK — APPROVED** [review]
Looks good to me. This is only for comp service. What about dynamic ones, can we have all this fanciness for them as well?

File: services/web/server/src/simcore_service_webserver/director_v2/_handlers.py (hunk 1 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/director_v2/_handlers.py]: pls keep it in mind.

**Review by pcrespov — APPROVED** [review]

File: services/static-webserver/client/source/class/osparc/data/model/Node.js [hunk 4]
```diff
@@ -967,7 +967,7 @@ qx.Class.define("osparc.data.model.Node", {
       osparc.data.Resources.fetch("studies", "startNode", params)
         .then(() => this.startDynamicService())
         .catch(err => {
-          if ("status" in err && err.status === 409) {
+          if ("status" in err && (err.status === 409 || err.status === 402)) {
```
Comment (sanderegg) [inline services/static-webserver/client/source/class/osparc/data/model/Node.js:970]: 402 = payment required... nice!

File: services/web/server/src/simcore_service_webserver/director_v2/_handlers.py [hunk 5]
```diff
@@ -69,83 +70,100 @@ class _ComputationStarted(BaseModel):
 @permission_required("services.pipeline.*")
 @permission_required("project.read")
 async def start_computation(request: web.Request) -> web.Response:
-    req_ctx = RequestContext.parse_obj(request)
-    computations = ComputationsApi(request.app)
+    # pylint: disable=too-many-statements
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/director_v2/_handlers.py:73]: I agree with pylint here... the maintainability of this method is not good.

**Review by sanderegg — APPROVED** [review]
Please correct the imports. otherwise it's looking very good!

File: services/web/server/src/simcore_service_webserver/director_v2/_handlers.py [hunk 6]
```diff
@@ -22,6 +22,7 @@
 )
 from simcore_service_webserver.db.plugin import get_database_engine
 from simcore_service_webserver.users.exceptions import UserDefaultWalletNotFoundError
+from simcore_service_webserver.wallets.errors import WalletNotEnoughCreditsError
```
[the remark this answers is not in the document]
Comment (matusdrobuliak66) [reply services/web/server/src/simcore_service_webserver/director_v2/_handlers.py:25]: yes you are right, thanks

File: services/web/server/src/simcore_service_webserver/director_v2/_handlers.py (hunk 5 above)
Comment (matusdrobuliak66) [reply services/web/server/src/simcore_service_webserver/director_v2/_handlers.py:73]: Yes, I have some ideas on how to refactor, will do, when the right time comes.

File: services/web/server/src/simcore_service_webserver/projects/_nodes_handlers.py [hunk 7]
```diff
@@ -44,6 +44,7 @@
 from servicelib.json_serialization import json_dumps
 from servicelib.mimetype_constants import MIMETYPE_APPLICATION_JSON
 from simcore_postgres_database.models.users import UserRole
+from simcore_service_webserver.wallets.errors import WalletNotEnoughCreditsError
```
[the remark this answers is not in the document]
Comment (matusdrobuliak66) [reply services/web/server/src/simcore_service_webserver/projects/_nodes_handlers.py:47]: thanks! what is the side effect of this? it seems it is working even without relative imports.

**Pull request merged**